File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/Python/122-Best-Time-to-Buy-and-Sell-Stock-II-Greedy.py
"""
Problem: 122. Best Time to Buy and Sell Stock II
Approach: Greedy

Whenever the next day's price is higher than today's price,
take that positive difference.

Because unlimited transactions are allowed, every positive
price increase can contribute to the maximum profit.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0

        for i in range(1, len(prices)):
            profit += max(0, prices[i] - prices[i - 1])

        return profit

File: Arrays/122-Best-Time-to-Buy-and-Sell-Stock-II/JavaScript/122-Best-Time-to-Buy-and-Sell-Stock-II-Greedy.js
/**
 * Problem: 122. Best Time to Buy and Sell Stock II
 * Approach: Greedy
 *
 * Whenever the next day's price is higher than today's price,
 * take that positive difference.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} prices
 * @return {number}
 */

var maxProfit = function(prices) {
    let profit = 0;

    for (let i = 1; i < prices.length; i++) {
        profit += Math.max(0, prices[i] - prices[i - 1]);
    }

    return profit;
};